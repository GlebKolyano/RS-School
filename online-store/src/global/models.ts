export interface IData {
  bicycles: IBicycle[];
}

export interface IBicycle {
  id: number;
  name: string;
  quantity: number;
  type: string;
  brand: string;
  speeds: number;
  color: string;
  weight: number;
  isPopular: boolean;
  inBasket: boolean;
  image: string;
  price: number;
}

export enum SortOptions {
  name_asc = 'name_asc',
  name_desc = 'name_desc',
  price_asc = 'price_asc',
  price_desc = 'price_desc',
  quantity_asc = 'quantity_asc',
  quantity_desc = 'quantity_desc'
}

export interface ISearchInitialState {
  searchValue: string;
}

export interface ISortInitialState {
  sortOption: string;
}

export interface IFilterByValueInitialState {
  filterByCompany: string[];
  filterByType: string[];
  filterByColor: string[];
  filterByPopular: boolean;
}

export type FilterByRangePayload = {
  min: number;
  max: number;
};
export interface IFilterByRangeInitialState {
  filterByQuantity: FilterByRangePayload;
  filterByPrice: FilterByRangePayload;
}

export interface IStorageFilters {
  filterByValue: IFilterByValueInitialState;
  filterByRange: IFilterByRangeInitialState;
  sort: ISortInitialState;
  search: ISearchInitialState;
}

export type StoragePropType =
  | IFilterByRangeInitialState
  | string
  | string[]
  | IFilterByValueInitialState
  | IBicycle[]
  | number[];

export type StorageReturnType = StoragePropType | null;

export interface ICartInitialState {
  itemsInCart: number[];
}
export interface IBicycleInitialState {
  bicycles: IBicycle[];
  isLoading: boolean;
  error: string;
}

type ModalObjType = {
  [id: string]: boolean;
};

export interface IModalInitialState {
  modalsID: ModalObjType;
}

export interface IStoreInitialState {
  filterByRangeReducer: IFilterByRangeInitialState;
  filterByValueReducer: IFilterByValueInitialState;
  bicycleReducer: IBicycleInitialState;
  cartReducer: ICartInitialState;
  modalReducer: IModalInitialState;
  searchReducer: ISearchInitialState;
  sortReducer: ISortInitialState;
}
