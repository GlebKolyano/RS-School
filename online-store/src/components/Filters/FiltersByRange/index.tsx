import React from 'react';
import { useAppDispatch } from '../../../hooks/reduxHooks';
import { FilterByRangePayload } from '../../../models/models';
import { setfilterByPrice, setfilterByQuantity } from '../../../store/reducers/filterByRange.slice';
import { getMinMaxValuesForRangeSlider } from '../../../store/helpers';
import Slider from '../../UI/Slider';

function FilterByRange() {
  const dispatch = useAppDispatch();
  const minMaxValues = getMinMaxValuesForRangeSlider();

  const handleChangePrice = (vl: FilterByRangePayload) => {
    dispatch(setfilterByPrice(vl));
  };

  const handleChangeQuantity = (vl: FilterByRangePayload) => {
    dispatch(setfilterByQuantity(vl));
  };
  return (
    <div>
      <p>Количество на складе:</p>
      <Slider
        minVl={minMaxValues.minQuantity}
        maxVl={minMaxValues.maxQuantity}
        onChange={handleChangeQuantity}
        name="filterByQuantity"
      />
      <p>По цене:</p>
      <Slider
        minVl={minMaxValues.minPrice}
        maxVl={minMaxValues.maxPrice}
        onChange={handleChangePrice}
        name="filterByPrice"
      />
    </div>
  );
}

export default FilterByRange;
