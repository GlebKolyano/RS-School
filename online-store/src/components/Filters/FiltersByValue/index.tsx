import React from 'react';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';
import { colorBicycle, companyBicycle, typeBicycle } from './constants';
import { Checkbox } from '../../UI/Checkbox';
import {
  setfilterByColor,
  setFilterByCompany,
  setfilterByPopular,
  setfilterByType
} from '../../../store/reducers/filterByValue.slice';

function FiltersByValue() {
  const { filterByType, filterByPopular, filterByColor, filterByCompany } = useAppSelector(
    (state) => state.filterByValueReducer
  );
  const dispatch = useAppDispatch();

  const handleChangeCompany = (target: HTMLInputElement) => {
    const { value: company } = target;
    dispatch(setFilterByCompany(company));
  };

  const handleChangeColor = (target: HTMLInputElement) => {
    const { value: color } = target;
    dispatch(setfilterByColor(color));
  };

  const handleChangeType = (target: HTMLInputElement) => {
    const { value: type } = target;
    dispatch(setfilterByType(type));
  };
  const handleChangeIsPopular = () => {
    dispatch(setfilterByPopular());
  };

  return (
    <div>
      <div>
        Производитель:
        <ul>
          {companyBicycle.map((item) => {
            return (
              <Checkbox
                key={item}
                name="companyBicycle"
                value={item}
                onChange={handleChangeCompany}
                state={filterByCompany}
              />
            );
          })}
        </ul>
      </div>
      <div>
        Тип велосипеда:
        <ul>
          {typeBicycle.map((item) => {
            return (
              <Checkbox
                key={item}
                name="typeBicycle"
                value={item}
                onChange={handleChangeType}
                state={filterByType}
              />
            );
          })}
        </ul>
      </div>
      <div>
        Цвет:
        <ul>
          {colorBicycle.map((item) => {
            return (
              <Checkbox
                key={item}
                name="colorBicycle"
                value={item}
                onChange={handleChangeColor}
                state={filterByColor}
              />
            );
          })}
        </ul>
      </div>
      <div>
        Только популярные:
        <ul>
          <li>
            <input
              type="checkbox"
              name="popularBicycle"
              onChange={handleChangeIsPopular}
              checked={filterByPopular}
            />{' '}
            Показать популярные
          </li>
        </ul>
      </div>
    </div>
  );
}

export default FiltersByValue;
