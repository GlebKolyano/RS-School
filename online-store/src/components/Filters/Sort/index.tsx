import React from 'react';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';
import { setSort } from '../../../store/reducers/sort.slice';
import { options } from './constants';

const Sort = () => {
  const { sortOption } = useAppSelector(({ sortReducer }) => sortReducer);
  const dispatch = useAppDispatch();

  function changeSortOption(value: string) {
    dispatch(setSort(value));
  }

  return (
    <div data-testid="sort">
      <select
        className="browser-default"
        onChange={(e) => changeSortOption(e.target.value)}
        defaultValue={sortOption}
      >
        {options.map((option) => {
          return (
            <option key={option.value} value={option.value}>
              {option.label}
            </option>
          );
        })}
      </select>
    </div>
  );
};

export default Sort;
