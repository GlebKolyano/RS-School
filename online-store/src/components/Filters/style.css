.filters {
  max-width: 20%;
  width: 100%;
  background-color: #262626;
  color: white;
}
