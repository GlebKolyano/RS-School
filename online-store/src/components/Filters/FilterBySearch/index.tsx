import React from 'react';
import './style.css';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';
import { setSearch } from '../../../store/reducers/filterBySearch.slice';

function FilterBySearch() {
  const dispatch = useAppDispatch();
  const { searchValue } = useAppSelector((state) => state.searchReducer);

  return (
    <div>
      <div className="input-field search">
        <label htmlFor="first_name">
          Поиск товаров:
          <input
            placeholder="Поиск"
            id="first_name"
            type="text"
            className="search__input validate"
            onChange={(e) => dispatch(setSearch(e.target.value))}
            value={searchValue}
            autoFocus
            autoComplete="off"
          />
        </label>
      </div>
    </div>
  );
}

export default FilterBySearch;
