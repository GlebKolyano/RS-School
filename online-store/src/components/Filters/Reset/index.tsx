import React from 'react';
import './style.css';
import LocaleStorage from '../../../helpers/LocaleStorage';
import { useAppDispatch } from '../../../hooks/reduxHooks';
import { updateStateFiltersByRange } from '../../../store/reducers/filterByRange.slice';
import { updateStateFilterBySearch } from '../../../store/reducers/filterBySearch.slice';
import { updateStateFiltersByValue } from '../../../store/reducers/filterByValue.slice';
import Button from '../../UI/Button';
import { fieldNamesForRestFilters } from './constants';
// import { Button, Card, Row, Col } from 'react-materialize';

function Reset() {
  const dispatch = useAppDispatch();
  const Storage = new LocaleStorage();

  const handleResetFilters = () => {
    fieldNamesForRestFilters.forEach((fieldName) => {
      dispatch(updateStateFiltersByRange());
      dispatch(updateStateFiltersByValue());
      dispatch(updateStateFilterBySearch());
      Storage.remove(fieldName);
    });
  };

  const handleResetSettings = () => {
    Storage.clear();
    window.location.reload();
  };

  return (
    <div className="reset">
      <Button
        onClick={handleResetFilters}
        text="Сбросить фильтры"
        type="button"
        iconName="filter_list"
      />
      <Button
        onClick={handleResetSettings}
        text="Сбросить настройки"
        type="button"
        iconName="settings"
      />
    </div>
  );
}

export default Reset;
