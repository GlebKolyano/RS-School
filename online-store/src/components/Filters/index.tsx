import React from 'react';
import './style.css';

function Filters() {
  return <div className="filters">Filters</div>;
}

export default Filters;
