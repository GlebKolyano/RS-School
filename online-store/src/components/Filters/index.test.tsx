import React from 'react';
import { render } from '@testing-library/react';
import Filters from '.';
import { renderWithRedux } from '../../tests/helpers/renderWithRedux';
import { IStoreInitialState } from '../../models/models';

const FakeInitialState: IStoreInitialState = {
  bicycleReducer: { bicycles: [], error: '', isLoading: false },
  cartReducer: { itemsInCart: [] },
  filterByRangeReducer: {
    filterByPrice: { min: 0, max: 100 },
    filterByQuantity: { min: 0, max: 100 }
  },
  filterByValueReducer: {
    filterByColor: [],
    filterByCompany: [],
    filterByPopular: false,
    filterByType: []
  },
  modalReducer: { modalsID: {} },
  searchReducer: { searchValue: '' },
  sortReducer: { sortOption: '' }
};

describe('filters', () => {
  describe('correctly renders', () => {
    test('iteself', () => {
      renderWithRedux(<Filters />, FakeInitialState);
    });
  });
});
