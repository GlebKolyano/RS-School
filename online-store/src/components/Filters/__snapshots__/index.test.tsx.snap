// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters correctly renders filters snapshot 1`] = `
<div
  class="filters"
  data-testid="filters"
>
  <div
    class="input-field search"
    data-testid="filter-by-search"
  >
    <label
      for="first_name"
    >
      Поиск товаров:
      <input
        autocomplete="off"
        class="search__input validate"
        id="first_name"
        placeholder="Поиск"
        type="text"
        value=""
      />
      <button
        class="search__clear"
        type="button"
      >
        <i
          class="material-icons"
        >
          clear
        </i>
      </button>
    </label>
  </div>
  <div
    data-testid="sort"
  >
    <select
      class="browser-default"
    >
      <option
        value="name_asc"
      >
        По названию, от А до Я
      </option>
      <option
        value="name_desc"
      >
        По названию, от Я до А
      </option>
      <option
        value="price_asc"
      >
        По цене, по возрастанию
      </option>
      <option
        value="price_desc"
      >
        По цене, по убыванию
      </option>
      <option
        value="quantity_asc"
      >
        По количеству, по возрастанию
      </option>
      <option
        value="quantity_desc"
      >
        По количеству, по убыванию
      </option>
    </select>
  </div>
  <div
    data-testid="filter-by-value"
  >
    <div>
      Производитель:
      <ul>
        <li
          data-testid="checkbox"
        >
          <label
            for="Denton"
          >
            <input
              class="filled-in"
              id="Denton"
              name="companyBicycle"
              type="checkbox"
              value="Denton"
            />
            <span>
              Denton
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="Eltreco"
          >
            <input
              class="filled-in"
              id="Eltreco"
              name="companyBicycle"
              type="checkbox"
              value="Eltreco"
            />
            <span>
              Eltreco
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="Ghost"
          >
            <input
              class="filled-in"
              id="Ghost"
              name="companyBicycle"
              type="checkbox"
              value="Ghost"
            />
            <span>
              Ghost
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="Giant"
          >
            <input
              class="filled-in"
              id="Giant"
              name="companyBicycle"
              type="checkbox"
              value="Giant"
            />
            <span>
              Giant
            </span>
          </label>
        </li>
      </ul>
    </div>
    <div>
      Тип велосипеда:
      <ul>
        <li
          data-testid="checkbox"
        >
          <label
            for="Горный"
          >
            <input
              class="filled-in"
              id="Горный"
              name="typeBicycle"
              type="checkbox"
              value="Горный"
            />
            <span>
              Горный
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="Электровелосипед"
          >
            <input
              class="filled-in"
              id="Электровелосипед"
              name="typeBicycle"
              type="checkbox"
              value="Электровелосипед"
            />
            <span>
              Электровелосипед
            </span>
          </label>
        </li>
      </ul>
    </div>
    <div>
      Цвет:
      <ul>
        <li
          data-testid="checkbox"
        >
          <label
            for="черный"
          >
            <input
              class="filled-in"
              id="черный"
              name="colorBicycle"
              type="checkbox"
              value="черный"
            />
            <span>
              черный
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="серый"
          >
            <input
              class="filled-in"
              id="серый"
              name="colorBicycle"
              type="checkbox"
              value="серый"
            />
            <span>
              серый
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="оранжевый"
          >
            <input
              class="filled-in"
              id="оранжевый"
              name="colorBicycle"
              type="checkbox"
              value="оранжевый"
            />
            <span>
              оранжевый
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="хаки"
          >
            <input
              class="filled-in"
              id="хаки"
              name="colorBicycle"
              type="checkbox"
              value="хаки"
            />
            <span>
              хаки
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="зеленый"
          >
            <input
              class="filled-in"
              id="зеленый"
              name="colorBicycle"
              type="checkbox"
              value="зеленый"
            />
            <span>
              зеленый
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="красный"
          >
            <input
              class="filled-in"
              id="красный"
              name="colorBicycle"
              type="checkbox"
              value="красный"
            />
            <span>
              красный
            </span>
          </label>
        </li>
        <li
          data-testid="checkbox"
        >
          <label
            for="синий"
          >
            <input
              class="filled-in"
              id="синий"
              name="colorBicycle"
              type="checkbox"
              value="синий"
            />
            <span>
              синий
            </span>
          </label>
        </li>
      </ul>
    </div>
    <div>
      Только популярные:
      <ul>
        <li>
          <label
            for="popularIdCheckbox"
          >
            <input
              class="filled-in"
              id="popularIdCheckbox"
              name="popularBicycle"
              type="checkbox"
            />
            <span>
              Показать популярные
            </span>
          </label>
        </li>
      </ul>
    </div>
  </div>
  <div
    data-testid="filter-by-range"
  >
    <div
      data-testid="slider"
    >
      <p>
        Количество на складе:
      </p>
      <span>
        от 
        0
         | 
      </span>
      <span>
        до 
        100
      </span>
      <br />
      <div
        class="rc-slider rc-slider-horizontal"
      >
        <div
          class="rc-slider-rail"
        />
        <div
          class="rc-slider-track rc-slider-track-1"
          style="left: 0%; width: 100%;"
        />
        <div
          class="rc-slider-step"
        />
        <div
          aria-disabled="false"
          aria-valuemax="99"
          aria-valuemin="5"
          aria-valuenow="5"
          class="rc-slider-handle rc-slider-handle-1"
          role="slider"
          style="left: 0%; transform: translateX(-50%);"
          tabindex="0"
        />
        <div
          aria-disabled="false"
          aria-valuemax="99"
          aria-valuemin="5"
          aria-valuenow="99"
          class="rc-slider-handle rc-slider-handle-2"
          role="slider"
          style="left: 100%; transform: translateX(-50%);"
          tabindex="0"
        />
      </div>
    </div>
    <div
      data-testid="slider"
    >
      <p>
        По цене:
      </p>
      <span>
        от 
        0
         | 
      </span>
      <span>
        до 
        100
      </span>
      <br />
      <div
        class="rc-slider rc-slider-horizontal"
      >
        <div
          class="rc-slider-rail"
        />
        <div
          class="rc-slider-track rc-slider-track-1"
          style="left: 0%; width: 0%;"
        />
        <div
          class="rc-slider-step"
        />
        <div
          aria-disabled="false"
          aria-valuemax="148899"
          aria-valuemin="13649"
          aria-valuenow="13649"
          class="rc-slider-handle rc-slider-handle-1"
          role="slider"
          style="left: 0%; transform: translateX(-50%);"
          tabindex="0"
        />
        <div
          aria-disabled="false"
          aria-valuemax="148899"
          aria-valuemin="13649"
          aria-valuenow="13649"
          class="rc-slider-handle rc-slider-handle-2"
          role="slider"
          style="left: 0%; transform: translateX(-50%);"
          tabindex="0"
        />
      </div>
    </div>
  </div>
  <div
    class="reset"
    data-testid="reset"
  >
    <button
      class="waves-effect waves-light btn button"
      type="button"
    >
      <i
        class="material-icons right"
      >
        filter_list
      </i>
      Сбросить фильтры
    </button>
    <button
      class="waves-effect waves-light btn button"
      type="button"
    >
      <i
        class="material-icons right"
      >
        settings
      </i>
      Сбросить настройки
    </button>
  </div>
</div>
`;
